File: dsaWithJava/tree/binarysearchtree/BSTIterative.java
package tree.binarysearchtree;

import java.util.Stack;

public class BSTIterative {
    Node root;

    public void insert(int x) {
        root = insert(root, x);
    }

    public void delete(int x) {
        root = delete(root, x);
    }

    public boolean search(int x) {
        return search(root, x);
    }

    public void display() {
        display(root);
    }

    // Time Complexity -> O(h)
    // Auxiliary Space -> O(1)
    private Node insert(Node root, int x) {
        Node newNode = new Node(x);
        Node parentNode = null, currentNode = root;
        while (currentNode != null) {
            parentNode = currentNode;
            if (currentNode.data > x)
                currentNode = currentNode.left;
            else if (currentNode.data < x)
                currentNode = currentNode.right;
            else
                return root;
        }
        if (parentNode == null)
            return newNode;
        if (parentNode.data > x)
            parentNode.left = newNode;
        else
            parentNode.right = newNode;
        return root;
    }

    // Time Complexity -> O(h)
    // Auxiliary Space -> O(1)
    private Node delete(Node root, int x) {
        if (root == null)
            return null;
        Node currentNode = root, parentNode = null;
        while (currentNode != null && currentNode.data != x) {
            parentNode = currentNode;
            if (currentNode.data > x)
                currentNode = currentNode.left;
            else
                currentNode = currentNode.right;
        }
        if (currentNode == null)
            return root;
        if (currentNode.left == null || currentNode.right == null) {
            Node newNode;
            if (currentNode.left == null)
                newNode = currentNode.right;
            else
                newNode = currentNode.left;
            if (parentNode == null)
                return newNode;
            if (currentNode == parentNode.left)
                parentNode.left = newNode;
            else
                parentNode.right = newNode;
        } else {
            Node inOrderParent = null, successor = currentNode.right;
            while (successor.left != null) {
                inOrderParent = successor;
                successor = successor.left;
            }
            if (inOrderParent != null)
                inOrderParent.left = successor.right;
            else
                currentNode.right = successor.right;
            currentNode.data = successor.data;
        }
        return root;
    }

    // Time Complexity -> O(h)
    // Auxiliary Space -> O(1)
    private static boolean search(Node root, int x) {
        while (root != null) {
            if (root.data == x)
                return true;
            else if (root.data > x)
                root = root.left;
            else
                root = root.right;
        }
        return false;
    }

    private void display(Node root) {
        if (root == null)
            return;
        Stack<Node> stk = new Stack<>();
        Node currentNode = root;
        System.out.print("[");
        while (currentNode != null || stk.isEmpty() == false) {
            while (currentNode != null) {
                stk.push(currentNode);
                currentNode = currentNode.left;
            }
            currentNode = stk.pop();
            System.out.print(currentNode.data + ", ");
            currentNode = currentNode.right;
        }
        System.out.println("\b\b]");
    }
}

File: Python/tree/binary_search_tree/bst_iterative.py
from node import Node

root = None


def insert(data):
    global root

    def __insert(root: Node, data) -> Node:
        new_node = Node(data)
        parent_node, current_node = None, root
        while current_node is not None:
            parent_node = current_node
            if current_node.data > data:
                current_node = current_node.left
            elif current_node.data < data:
                current_node = current_node.right
            else:
                return root
        if parent_node is None:
            return new_node
        if parent_node.data > data:
            parent_node.left = new_node
        else:
            parent_node.right = new_node
        return root

    root = __insert(root, data)


def delete(data):
    global root

    def __delete(root: Node, data) -> Node:
        if root is None:
            return None
        current_node = root
        parent_node = None
        while current_node is not None and current_node.data != data:
            parent_node = current_node
            if current_node.data > data:
                current_node = current_node.left
            else:
                current_node = current_node.right
        if current_node is None:
            return root
        if current_node.left is None or current_node.right is None:
            new_node = None
            if current_node.left is None:
                new_node = current_node.right
            else:
                new_node = current_node.left
            if parent_node is None:
                return new_node
            if current_node == parent_node.left:
                parent_node.left = new_node
            else:
                parent_node.right = new_node
        else:
            in_order_parent = None
            successor = current_node.right
            while successor.left is not None:
                in_order_parent = successor
                successor = successor.left
            if in_order_parent is not None:
                in_order_parent.left = successor.right
            else:
                current_node.right = successor.right
            current_node.data = successor.data
        return root

    root = __delete(root, data)


def search(data) -> bool:
    global root

    def __search(root: Node, data) -> bool:
        while root is not None:
            if root.data == data:
                return True
            if root.data > data:
                root = root.left
            else:
                root = root.right
        return False

    return __search(root, data)


def display():
    global root

    def __display(root: Node):
        stack = []
        current_node = root
        print("[", end="")
        while current_node is not None or len(stack) > 0:
            while current_node is not None:
                stack.append(current_node)
                current_node = current_node.left
            current_node = stack.pop()
            print(current_node.data, end=", ")
            current_node = current_node.right
        print("\b\b]")

    __display(root)
